program/src/guards/freeze_sol_payment.rs

```diff
@@ -0,0 +1,493 @@
+use super::*;
+
+use anchor_lang::AccountsClose;
+use mpl_candy_machine_core::CandyMachine;
+use mpl_token_metadata::instruction::{freeze_delegated_account, thaw_delegated_account};
+use solana_program::{
+    program::{invoke, invoke_signed},
+    program_pack::Pack,
+    system_instruction,
+};
+use spl_token::{
+    instruction::{approve, revoke},
+    state::Account as TokenAccount,
+};
+
+use crate::{
+    errors::CandyGuardError,
+    utils::{assert_is_ata, assert_keys_equal, cmp_pubkeys},
+};
+
+/// Guard that charges an amount in SOL (lamports) for the mint.
+///
+/// List of accounts required:
+///
+///   0. `[writable]` Freeze PDA to receive the funds (seeds `["freeze_sol_payment",
+///           destination pubkey, candy guard pubkey, candy machine pubkey]`).
+///   1. `[]` NFT mint associated account (seeds `[payer pubkey, nft mint pubkey]`)
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
+pub struct FreezeSolPayment {
+    pub lamports: u64,
+    pub destination: Pubkey,
+}
+
+impl Guard for FreezeSolPayment {
+    fn size() -> usize {
+        8    // lamports
+        + 32 // destination
+    }
+
+    fn mask() -> u64 {
+        0b1u64 << 16
+    }
+
+    /// Instructions to interact with the freeze feature:
+    ///
+    ///  * initialize
+    ///  * thawn
+    ///  * unlock funds
+    fn instruction<'info>(
+        ctx: &Context<'_, '_, '_, 'info, Route<'info>>,
+        guard_set: &GuardSet,
+        data: Vec<u8>,
+    ) -> Result<()> {
+        // determines the instruction to execute
+        let instruction: FreezeInstruction =
+            if let Ok(instruction) = FreezeInstruction::try_from_slice(&data[0..1]) {
+                instruction
+            } else {
+                return err!(CandyGuardError::MissingFreezeInstruction);
+            };
+
+        let freeze_guard = if let Some(freeze_guard) = &guard_set.freeze_sol_payment {
+            freeze_guard
+        } else {
+            return err!(CandyGuardError::FreezeGuardNotEnabled);
+        };
+
+        match instruction {
+            // Initializes the freese escrow PDA.
+            //
+            // List of accounts required:
+            //
+            //   0. `[writable]` Freeze PDA to receive the funds (seeds `["freeze_sol_payment",
+            //                   destination pubkey, candy guard pubkey, candy machine pubkey]`).
+            //   1. `[signer]` Candy Guard authority.
+            //   2. `[]` System program account.
+            FreezeInstruction::Initialize => {
+                msg!("FreezeSolPayment: initialize");
+
+                let candy_guard_key = &ctx.accounts.candy_guard.key();
+                let candy_machine_key = &ctx.accounts.candy_machine.key();
+
+                let seeds = [
+                    b"freeze_sol_payment".as_ref(),
```

Comment:
`freeze_sol_payment` should probably be a constant. I'd impl a const called `PREFIX` on `FreezeEscrow`. It looks like a lot of other PDAs might not use constants in guard either, so perhaps this should be a separate issue to potentially update all the PDAs to reference constants as the seed prefix.

Reply:
Done.